EIPS/eip-1559.md

```diff
@@ -34,48 +38,168 @@ An important aspect of this fee system is that miners only get to keep the gas p
-    0
-  else
-    EIP1559_INITIAL_GAS_TARGET + (BLOCK_GAS_TARGET / 2) * (CURRENT_BLOCK_NUMBER - INITIAL_FORK_BLOCK_NUMBER) / MIGRATION_DURATION_IN_BLOCKS
-  ```
-* `LEGACY_GAS_LIMIT`: `BLOCK_GAS_TARGET - EIP1559_GAS_TARGET`. The maximum amount of gas legacy transactions can use in a given block.
-* `EIP1559_GAS_LIMIT`: `EIP1559_GAS_TARGET * 2`. The maximum amount of gas EIP-1559 transactions can use in a given block.
-* `BASE_FEE`: 16th item in the block header. Represents the amount of attoeth burned for every unit of gas a transaction uses.
-* `BASE_FEE_MAX_CHANGE_DENOMINATOR`: `8`
-* `BASE_FEE_MAX_CHANGE`: `CURRENT_BLOCK.parent.BASE_FEE / BASE_FEE_MAX_CHANGE_DENOMINATOR`. The maximum amount the `BASE_FEE` can change between blocks (either up or down).
-* `INITIAL_BASE_FEE` : 0.000000001 ETH (1 nanoeth)
-* `BASE_FEE_CHANGE`: `CURRENT_BLOCK.parent.BASE_FEE`. The amount that the BASE_FEE will increase in `CURRENT_BLOCK` relative to `CURRENT_BLOCK.parent`.
-
-### Process
-* At `block.number == INITIAL_FORK_BLOCK_NUMBER` we set `BASE_FEE = INITIAL_BASE_FEE`
-* `BASE_FEE` is set as follows
-  * Let `delta = block.gas_used - TARGET_GASUSED` (possibly negative).
-  * Set `BASE_FEE = PARENT_BASE_FEE + PARENT_BASE_FEE * delta // TARGET_GASUSED // BASE_FEE_MAX_CHANGE_DENOMINATOR`
-  * Clamp the resulting `BASE_FEE` inside of the allowable bounds if needed, where a valid `BASE_FEE` is one such that `abs(BASE_FEE - PARENT_BASE_FEE) <= max(1, PARENT_BASE_FEE // BASE_FEE_MAX_CHANGE_DENOMINATOR)`
-* EIP-1559 transactions are encoded as `rlp([nonce, gasLimit, to, value, data, gasPremium, feeCap, v, r, s])` where `v,r,s` is a signature of `rlp([nonce, gasLimit, to, value, data, gasPremium, feeCap])`.
-* During the transition phase, legacy transactions (`rlp([nonce, gasPrice, gasLimit, to, value, data, v, r, s])`) will continue to be valid and can be included as part of the `LEGACY_GAS_LIMIT` block of transactions.
-* To produce an EIP1559 transactions, the new `GAS_PREMIUM` and `FEECAP` fields are set as follows:
-  * `GAS_PREMIUM` serves as a "tip" to the miner
-  * `FEECAP` serves as the absolute maximum that the transaction sender is willing to pay
-* During transaction execution, for EIP1559 transactions we calculate the cost to the `tx.origin` and the gain to the `block.coinbase` as follows:
-  * Set `GASPRICE = min(BASE_FEE + tx.GasPremium, tx.fee_cap)`
-  * Let `GASUSED` be the gas used during the transaction execution/state transition
-  * The `tx.origin` initially pays `GASPRICE * tx.gasLimit`, and gets refunded `GASPRICE * (tx.gasLimit - GASUSED)`
-  * The `block.coinbase` gains `(GASPRICE - BASE_FEE) * GASUSED`.
-    * If `GASPRICE < BASE_FEE` (due to the `FEECAP`), this means that the `block.coinbase` _loses_ funds from this operation; in this case, we check that the post-balance is non-negative and throw an exception if it is negative.
+*Note: `//` in Python is integer division with rounding toward infinity (not truncation).*
```

Comment:
Should further clarify here that Python `//` rounds negative results to negative infinity and positive results to 0.

Reply:
Fixed this to indicate that Python `//` rounds toward `0`.

For those not following along in Discord, here is a synopsis of the rounding issue:

Besu originally implemented truncation and Geth implemented round toward negative infinity.  Upon noticing consensus failure Besu changed to match Geth.  However, every modern language (C++11, Java, C#, JavaScript, Python, Go, Rust) uses truncation for integer division (round toward 0) *including Go* (which Geth is written in) and the only reason Geth initially rounded toward negative infinity is because the big number library that Geth uses rounds toward negative infinity (for reasons that I'm not familiar with).  As far as we know, this is the first time integer division with negative numbers has been part of consensus, so this is the first time this has come up.

Since every language rounds toward 0 (many of which natively support large integers), I argue that this spec should specify that we round toward 0.

@i-norden argues (please correct me if I'm misrepresenting your position) that since Geth is the oldest and most calcified codebase, we should do what is easiest for Geth since it doesn't matter *functionally* which direction we round (at least in this case).

Reply:
Python `//` rounds to 0 only when the result is positive, it rounds to negative infinity when it is negative. I have confirmed this in a Python interpreter. Both besu and geth are conformed to this, I think we should remain this way.